### ISSUE TITLE
Error on new cluster boot: secret "moco-mycluster-grpc" not found

### ISSUE THREAD

**Issue by mplewis** [cybozu-go/moco#426]
**Describe the bug**

I am trying to start a new Moco cluster. However, the pod does not start successfully, and I see the following error event on the pod:

```
MountVolume.SetUp failed for volume "grpc-cert" : secret "moco-mycluster-grpc" not found
```

My cluster config is as follows:

```yaml
apiVersion: moco.cybozu.com/v1beta2
kind: MySQLCluster
metadata:
  name: mycluster
  namespace: default
spec:
  backupPolicyName: daily
  podTemplate:
    metadata: {}
    spec:
      containers:
      - image: quay.io/cybozu/mysql:8.0.28
        name: mysqld
        resources:
          limits:
            cpu: "1"
            memory: 1Gi
          requests:
            cpu: 100m
            memory: 128Mi
  replicas: 1
  volumeClaimTemplates:
  - metadata:
      name: mysql-data
    spec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 1Gi
```

**Environments**

- Version: `version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.14", GitCommit:"0f77da5bd4809927e15d1658fb4aa8f13ad890a5", GitTreeState:"clean", BuildDate:"2022-06-15T14:11:36Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}`
- OS: Linux (DigitalOcean Managed Kubernetes Cluster)

**To Reproduce**

* Start a [managed Kubernetes cluster on DigitalOcean](https://www.digitalocean.com/products/kubernetes)
* Install Cert-Manager using Helm release 1.8.2
* Install Moco using Helm release 0.2.3
* Create the cluster from the YAML above

**Expected behavior**

The cluster creates without error.

Please let me know if you need any additional info to help debug this.

**Comment by zoetrope** [comment]
The secret should be created before the StatefulSet.

Please run the following command and tell me if you have the Secret.

```
kubectl get secret moco-mycluster-grpc
```

Could you provide moco-controller's logs of when you created the cluster?

**Comment by mplewis** [comment]
Yes, this seems to be an issue: the secret isn't created when the StatefulSet comes online:

```
$ kubectl get secret moco-mycluster-grpc
Error from server (NotFound): secrets "moco-mycluster-grpc" not found
```

In the logs, I notice that there are three operations occurring which relate to secrets:

* created controller Secret (`mysql-default.mycluster`)
* reconciled user Secret (`moco-mycluster`)
* reconciled my.cnf Secret (`moco-my-cnf-mycluster`)

Here are the full logs from after I create the fresh cluster:

```json
{"level":"info","ts":1657925552.3789482,"logger":"controller.mysqlcluster","msg":"created controller Secret","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","secretName":"mysql-default.mycluster"}
{"level":"info","ts":1657925552.392066,"logger":"controller.mysqlcluster","msg":"reconciled user Secret","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","secretName":"moco-mycluster"}
{"level":"info","ts":1657925552.406889,"logger":"controller.mysqlcluster","msg":"reconciled my.cnf Secret","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","secretName":"moco-my-cnf-mycluster"}
{"level":"info","ts":1657925552.5299957,"logger":"controller.mysqlcluster","msg":"reconciled my.cnf ConfigMap","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","configMapName":"moco-mycluster.953412cc"}
{"level":"info","ts":1657925552.5386188,"logger":"controller.mysqlcluster","msg":"reconciled ConfigMap for slow logs","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","configMapName":"moco-slow-log-agent-config-mycluster"}
{"level":"info","ts":1657925552.5647366,"logger":"controller.mysqlcluster","msg":"reconciled ServiceAccount","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","serviceAccountName":"moco-mycluster"}
{"level":"info","ts":1657925552.5794935,"logger":"controller.mysqlcluster","msg":"reconciled Service","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","serviceName":"moco-mycluster"}
{"level":"info","ts":1657925552.60589,"logger":"controller.mysqlcluster","msg":"reconciled Service","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","serviceName":"moco-mycluster-primary"}
{"level":"info","ts":1657925552.6343389,"logger":"controller.mysqlcluster","msg":"reconciled Service","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","serviceName":"moco-mycluster-replica"}
{"level":"info","ts":1657925552.658057,"logger":"controller.mysqlcluster","msg":"reconciled StatefulSet","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","statefulSetName":"moco-mycluster"}
{"level":"info","ts":1657925552.6797945,"logger":"controller.mysqlcluster","msg":"reconciled CronJob for backup","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","cronJobName":"moco-backup-mycluster"}
{"level":"info","ts":1657925552.764195,"logger":"controller.mysqlcluster","msg":"reconciled Role for backup","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","roleName":"moco-backup-mycluster"}
{"level":"info","ts":1657925552.7883966,"logger":"controller.mysqlcluster","msg":"reconciled RoleBinding for backup","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","roleBindingName":"moco-backup-mycluster"}
{"level":"info","ts":1657925552.8288941,"logger":"controller.mysqlcluster","msg":"reconciled Service","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","serviceName":"moco-mycluster"}
{"level":"info","ts":1657925552.833296,"logger":"cluster-manager.default/mycluster","msg":"update the status information"}
{"level":"info","ts":1657925552.8453188,"logger":"controller.mysqlcluster","msg":"reconciled Service","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","serviceName":"moco-mycluster-primary"}
{"level":"info","ts":1657925552.8542955,"logger":"controller.mysqlcluster","msg":"reconciled Service","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","serviceName":"moco-mycluster-replica"}
{"level":"info","ts":1657925552.862228,"logger":"cluster-manager.default/mycluster","msg":"cluster state is Lost"}
{"level":"info","ts":1657925552.8741777,"logger":"controller.mysqlcluster","msg":"reconciled CronJob for backup","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","cronJobName":"moco-backup-mycluster"}
{"level":"info","ts":1657925552.893208,"logger":"controller.mysqlcluster","msg":"reconciled Service","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","serviceName":"moco-mycluster"}
{"level":"info","ts":1657925552.8963215,"logger":"cluster-manager.default/mycluster","msg":"cluster state is Lost"}
{"level":"info","ts":1657925552.9022868,"logger":"controller.mysqlcluster","msg":"reconciled Service","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","serviceName":"moco-mycluster-primary"}
{"level":"info","ts":1657925552.941502,"logger":"controller.mysqlcluster","msg":"reconciled Service","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","serviceName":"moco-mycluster-replica"}
{"level":"info","ts":1657925552.979733,"logger":"controller.mysqlcluster","msg":"reconciled CronJob for backup","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","cronJobName":"moco-backup-mycluster"}
{"level":"info","ts":1657925553.0171769,"logger":"cluster-manager.default/mycluster","msg":"cluster state is Lost"}
{"level":"info","ts":1657925553.2946203,"logger":"controller.mysqlcluster","msg":"reconciled Service","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","serviceName":"moco-mycluster"}
{"level":"info","ts":1657925553.3090465,"logger":"controller.mysqlcluster","msg":"reconciled Service","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","serviceName":"moco-mycluster-primary"}
{"level":"info","ts":1657925553.3295047,"logger":"controller.mysqlcluster","msg":"reconciled Service","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","serviceName":"moco-mycluster-replica"}
{"level":"info","ts":1657925553.3560166,"logger":"controller.mysqlcluster","msg":"reconciled CronJob for backup","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","cronJobName":"moco-backup-mycluster"}
{"level":"info","ts":1657925553.4154,"logger":"cluster-manager.default/mycluster","msg":"cluster state is Lost"}
{"level":"info","ts":1657925553.4170942,"logger":"controller.mysqlcluster","msg":"reconciled Service","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","serviceName":"moco-mycluster"}
{"level":"info","ts":1657925553.490396,"logger":"controller.mysqlcluster","msg":"reconciled Service","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","serviceName":"moco-mycluster-primary"}
{"level":"info","ts":1657925553.5258667,"logger":"controller.mysqlcluster","msg":"reconciled Service","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","serviceName":"moco-mycluster-replica"}
{"level":"info","ts":1657925553.54645,"logger":"controller.mysqlcluster","msg":"reconciled CronJob for backup","reconciler group":"moco.cybozu.com","reconciler kind":"MySQLCluster","name":"mycluster","namespace":"default","cronJobName":"moco-backup-mycluster"}
{"level":"info","ts":1657925553.564164,"logger":"cluster-manager.default/mycluster","msg":"cluster state is Lost"}
```

Not sure if this helps or muddies the waters, but I did not see this specific issue when I booted a test cluster in my local kind cluster.

**Comment by mplewis** [comment]
Just confirming that I am using Moco Helm chart version 0.2.3 from ghcr.io/cybozu-go/moco.

Reviewing [the reconciler](https://github.com/cybozu-go/moco/blob/ae0c853b7ac893fd6c3bfbb819f2b57d7d5ee97e/controllers/mysqlcluster_controller.go#L201-L220=), it seems clear to me that I should see logs in this order:

* created controller Secret
* reconciled user Secret
* reconciled my.cnf Secret
* **reconciled gRPC secret**
* reconciled my.cnf ConfigMap

I wonder if the creation of this secret is falling through silently in the [client.IgnoreNotFound case handler](https://github.com/cybozu-go/moco/blob/ae0c853b7ac893fd6c3bfbb819f2b57d7d5ee97e/controllers/certificate.go#L83-L86=)?

**Comment by zoetrope** [comment]
Thanks for your information.

"moco-mycluster-grpc" is generated by cert-manager.
Could you please check the status of the certificate as follows?

```
kubectl describe certificate moco-agent-default.mycluster
```

**Comment by mplewis** [comment]
I see, thanks for clarifying.

Here is the result of `kubectl describe certificate moco-agent-default.mycluster`:

```
Name:         moco-agent-default.mycluster
Namespace:    default
Labels:       app.kubernetes.io/created-by=moco
              app.kubernetes.io/instance=mycluster
              app.kubernetes.io/instance-namespace=default
              app.kubernetes.io/name=mysql
Annotations:  <none>
API Version:  cert-manager.io/v1
Kind:         Certificate

...snip...

Status:
  Conditions:
    Last Transition Time:        2022-07-15T23:32:45Z
    Message:                     Issuing certificate as Secret does not exist
    Observed Generation:         1
    Reason:                      DoesNotExist
    Status:                      False
    Type:                        Ready
    Last Transition Time:        2022-07-15T23:32:45Z
    Message:                     Issuing certificate as Secret does not exist
    Observed Generation:         1
    Reason:                      DoesNotExist
    Status:                      True
    Type:                        Issuing
  Next Private Key Secret Name:  moco-agent-default.mycluster-wrvbx
Events:                          <none>
```

Investigating deeper with `moco-agent-default.mycluster-6g87q`:

```
...snip...

Status:
  Conditions:
    Last Transition Time:  2022-07-15T23:32:46Z
    Message:               Certificate request has been approved by cert-manager.io
    Reason:                cert-manager.io
    Status:                True
    Type:                  Approved
    Last Transition Time:  2022-07-15T23:32:46Z
    Message:               Referenced "Issuer" not found: issuer.cert-manager.io "moco-grpc-issuer" not found
    Reason:                Pending
    Status:                False
    Type:                  Ready
```

I do not see the Issuer `moco-grpc-issuer`, but here's what I have for `kubectl describe issuer moco-8a698ec5-grpc-issuer`:

```
Name:         moco-8a698ec5-grpc-issuer
Namespace:    default
Labels:       app.kubernetes.io/managed-by=Helm
              app.kubernetes.io/name=moco
              app.kubernetes.io/version=0.12.1
              helm.sh/chart=moco-0.2.3
Annotations:  meta.helm.sh/release-name: moco-8a698ec5
              meta.helm.sh/release-namespace: default
API Version:  cert-manager.io/v1
Kind:         Issuer

...snip...

Spec:
  Ca:
    Secret Name:  moco-grpc-ca
Status:
  Conditions:
    Last Transition Time:  2022-07-14T20:31:19Z
    Message:               Signing CA verified
    Observed Generation:   1
    Reason:                KeyPairVerified
    Status:                True
    Type:                  Ready
```

I notice that [`certificate.yaml`](https://github.com/cybozu-go/moco/blob/main/charts/moco/templates/certificate.yaml#L12=) and [`issuer.yaml`](https://github.com/cybozu-go/moco/blob/main/charts/moco/templates/issuer.yaml#L4=) use `name: {{ template "moco.fullname" . }}-grpc-issuer`, but [`certificate_tmpl.yaml`](https://github.com/cybozu-go/moco/blob/ae0c853b7ac893fd6c3bfbb819f2b57d7d5ee97e/controllers/certificate_tmpl.yaml#L16=) uses `moco-grpc-issuer`.

I wonder if this bug is exposed by `moco.fullname` being set to `moco-8a698ec5` and not `moco` in my Helm release?

**Comment by ymmt2005** [comment]
Looks like a bug in the helm chart.

**Comment by zoetrope** [comment]
@mplewis 
Thanks!
I fixed it: #432
